### PR TITLE
CSS-5249: Add OpenFGA relation

### PR DESCRIPTION
**Pull request by kelkawi-a**
This PR introduces a relation with the [OpenFGA](https://charmhub.io/openfga-k8s) charm. I have not updated the tutorial/instructions yet because all this does for the time being is inject the necessary OpenFGA environment variables into our workload container where they will not be used for the time being.

There is currently a WIP [PR](https://github.com/canonical/charmed-temporal-image/pull/3) that introduces a custom image of the Temporal server, which will use these environment variables to connect to the OpenFGA store and implement the necessary authorization logic. Once merged, I will update the necessary documentation to reflect on how to enable auth in Temporal.

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -228,6 +248,15 @@ def _validate(self):
         if not self._state.schema_ready:
             raise ValueError("admin:temporal relation: schema is not ready")
 
+        if self.config["auth-enabled"]:
+            if not self._state.openfga:
+                raise ValueError("openfga:temporal relation not ready")
+            missing_params = self._check_missing_openfga_params()
+            if len(missing_params) > 0:
+                raise ValueError(f"openfga:missing parameters {missing_params!r}")
```
Comment (georgeheshmet) [inline src/charm.py:256]: Is the "r" in the end of the {missing_params!r} intended?

**Review by georgeheshmet — APPROVED** [review]
LGTM :+1:

File: src/relations/openfga.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        url = f"{openfga_scheme}://{openfga_address}:{openfga_port}/stores/{openfga_store_id}/authorization-models"
+        headers = {"Content-Type": "application/json"}
+        if openfga_token:
+            headers["Authorization"] = f"Bearer {openfga_token}"
+
+        # do the post request
+        logger.info(f"posting to {url}, with headers {headers}")
+        try:
+            response = requests.post(url, json=model_json, headers=headers, timeout=10)
+        except requests.HTTPError as error:
```
Comment (maabujayyab) [inline src/relations/openfga.py:118]: Are we just interested in `HTTPError`? you might want to consider catching as well `RequestException` for other errors.

File: src/relations/openfga.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+            event.defer()
+            return
+
+        if self.charm.unit.is_leader():
+            self.charm._state.openfga = None
+            self.charm._update(event)
+
+    @log_event_handler(logger)
+    def _on_create_authorization_model_action(self, event):
+        """Handle OpenFGA create authorization model action.
```
Comment (maabujayyab) [inline src/relations/openfga.py:86]: Not a fan of this function's length, but it's up to you to refactor it to more readable sections.

**Review by maabujayyab — CHANGES_REQUESTED** [review]

File: actions.yaml [hunk 4]
```diff
@@ -3,3 +3,14 @@
 
 restart:
   description: Restart the Temporal server.
+
+create-authorization-model:
```
Comment (kian99) [inline actions.yaml:7]: If it helps, I struggled to figure out how to get a big yaml file into the parameter. The following did the trick
`juju run-action <app>/leader create-authorization-model --params auth_model.yaml --wait`
and then `auth_model.yaml ` looked like the following
```
model: >
  {
    "type_definitions": [],
    ...
  }
```

File: tests/unit/test_charm.py [hunk 5]
```diff
@@ -312,6 +312,92 @@ def test_ingress(self):
...
+        harness.update_config({"auth-enabled": True})
+
+        secret_id = harness.add_model_secret("temporal-k8s", {"token": "openfga_token"})
+        event = make_openfga_store_created_event(secret_id)
+        harness.charm.openfga_relation._on_openfga_store_created(event)
+
+        # database relation ready but no openfga store set up
+        self.assertEqual(harness.model.unit.status, BlockedStatus("missing openfga authorization model"))
+
+    def test_authorization_ready(self):
```
Comment (kian99) [inline tests/unit/test_charm.py:339]: I like all the various tests.

**Review by kian99 — APPROVED** [review]
lgtm

File: actions.yaml (hunk 4 above)
Comment (kelkawi-a) [reply actions.yaml:7]: I've also found this, which I will include in the deployment instructions in a future PR:
```
juju run <app>/leader create-authorization-model  model="$(<temporal_auth_model.json)" --string-args=true
```

File: tests/unit/test_charm.py (hunk 5 above)
Comment (kelkawi-a) [reply tests/unit/test_charm.py:339]: Thanks!

File: src/charm.py (hunk 1 above)
Comment (kelkawi-a) [reply src/charm.py:256]: Yes, it's used in f-strings to convert the value to its representation string (i.e. `value` would be output as `'value'`)

File: src/relations/openfga.py (hunk 3 above)
Comment (kelkawi-a) [reply src/relations/openfga.py:87]: I've refactored this into smaller, more readable functions.

**Review by mertalpt — COMMENTED** [review]
LGTM. However, I do not feel knowledgeable enough to approve a change that includes both Temporal and OpenFGA. So, this will be just a comment. Ping me if you need the approval and I will take another look.

File: src/charm.py [hunk 6]
```diff
@@ -205,6 +211,19 @@ def _on_restart_action(self, event):
...
+    def _check_missing_openfga_params(self):
+        """Validate that all OpenFGA required properties were extracted.
+
+        Returns:
+            list: List of OpenFGA parameters that are not set in state.
+        """
+        missing_params = []
+        required_openfga_keys = ["store_id", "address", "port", "scheme", "token"]
+        for key in required_openfga_keys:
+            if self._state.openfga.get(key) is None:
```
[the remark this answers is not in the document]
Comment (kelkawi-a) [reply src/charm.py:223]: No, it must have a value.

**Review by maabujayyab — APPROVED** [review]
LGTM

**Pull request merged**